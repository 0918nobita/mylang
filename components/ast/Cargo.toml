[package]
name = "ast"
version = "0.1.0"
edition = "2021"

[dependencies]
anyhow = "1.0.53"
