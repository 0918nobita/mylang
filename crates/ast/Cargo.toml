[package]
name = "mylang_ast"
version = "0.1.0"
edition = "2021"

[dependencies]
mylang_token = { path = "../token" }
serde = { version = "1.0.167", features = ["derive"] }
