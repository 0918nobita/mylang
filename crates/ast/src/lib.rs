mod expr;
mod stmt;

pub use expr::Expr;
pub use stmt::Stmt;
