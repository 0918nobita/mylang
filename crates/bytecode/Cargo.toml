[package]
name = "mylang_bytecode"
version = "0.1.0"
edition = "2021"

[dependencies]
serde = { version = "1.0.156", features = ["derive"] }
