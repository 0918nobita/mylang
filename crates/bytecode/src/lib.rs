//! バイトコードの定義

use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub enum Inst {
    I32Const(i32),
    I32Add,
    StrConst(String),
    PrintI32,
    PrintStr,
}
