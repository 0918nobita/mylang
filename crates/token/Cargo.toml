[package]
name = "mylang_token"
version = "0.1.0"
edition = "2021"

[dependencies]
serde = { version = "1.0.143", features = ["derive"] }
