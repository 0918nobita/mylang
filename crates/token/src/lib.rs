//! 字句解析で取り出されるトークンとその付帯情報の定義

mod locatable;
mod pos;
mod range;

use serde::{Deserialize, Serialize};

pub use locatable::Locatable;
pub use pos::Pos;
pub use range::Range;

/// [`Pos`] を簡単に生成するためのマクロ
#[macro_export]
macro_rules! pos {
    ($line:expr ; $char:expr) => {
        mylang_token::Pos::new($line, $char)
    };
}

/// [`Range`] を簡単に生成するためのマクロ
#[macro_export]
macro_rules! range {
    ($start_line:expr ; $start_char:expr, $end_line:expr ; $end_char:expr) => {
        mylang_token::Range::new(
            mylang_token::pos!($start_line;$start_char),
            mylang_token::pos!($end_line;$end_char),
        )
    };
    ($start:expr, $end:expr) => {
        mylang_token::Range::new($start, $end)
    };
}

#[derive(Debug, Serialize, Deserialize)]
pub enum KeywordKind {
    Let,
    In,
    PrintI32,
    PrintStr,
}

impl KeywordKind {
    pub fn parse(str: &str) -> Option<Self> {
        match str {
            "let" => Some(Self::Let),
            "in" => Some(Self::In),
            "print_int" => Some(Self::PrintI32),
            "print_str" => Some(Self::PrintStr),
            _ => None,
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub enum Token {
    I32(Range, i32),
    AddOp(Pos),
    Str(Range, String),
    Keyword(Range, KeywordKind),
    Ident(Range, String),
    Newline(Pos),
}

impl Locatable for Token {
    fn locate(&self) -> Range {
        match self {
            Self::I32(r, _) | Self::Str(r, _) | Self::Keyword(r, _) | Self::Ident(r, _) => {
                r.clone()
            }

            Self::AddOp(p) | Self::Newline(p) => p.clone().into(),
        }
    }
}
