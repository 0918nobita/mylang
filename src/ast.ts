export type NullLiteral = Readonly<{
    type: 'nullLiteral';
}>;

export type BooleanLiteral = Readonly<{
    type: 'booleanLiteral';
    value: boolean;
}>;

export type NumberLiteral = Readonly<{
    type: 'numberLiteral';
    value: number;
}>;

export type StringLiteral = Readonly<{
    type: 'stringLiteral';
    value: string;
}>;

export type Literal =
    | NullLiteral
    | BooleanLiteral
    | NumberLiteral
    | StringLiteral;

export type Identifier = Readonly<{
    type: 'identifier';
    name: string;
}>;

export type BinaryExpression = Readonly<{
    type: 'binaryExpression';
    operator: '+' | '*';
    left: Expression;
    right: Expression;
}>;

export type CallExpression = Readonly<{
    type: 'callExpression';
    callee: Expression;
    arguments: Expression[];
    optional: boolean;
}>;

export type MemberExpression = Readonly<{
    type: 'memberExpression';
    object: Expression;
    property: Expression;
    computed: boolean;
    optional: boolean;
}>;

export type Expression =
    | Literal
    | Identifier
    | BinaryExpression
    | CallExpression
    | MemberExpression;
