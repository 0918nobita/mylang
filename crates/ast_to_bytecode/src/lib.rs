use ast::{Expr, Stmt};
use bytecode::Inst;

fn expr_to_bytecode(expr: &Expr) -> Vec<Inst> {
    match expr {
        Expr::I32Lit(_, i) => vec![Inst::I32Const(*i)],
        Expr::Add(lhs, rhs) => {
            let mut insts = expr_to_bytecode(lhs);
            insts.extend(expr_to_bytecode(rhs));
            insts.push(Inst::I32Add);
            insts
        }
        Expr::StrLit(_, s) => vec![Inst::StrConst(s.clone())],
    }
}

fn stmt_to_bytecode(stmt: &Stmt) -> Vec<Inst> {
    match stmt {
        Stmt::PrintI32(_, ref expr) => {
            let mut insts = expr_to_bytecode(expr);
            insts.push(Inst::PrintI32);
            insts
        }
        Stmt::PrintStr(_, ref expr) => {
            let mut insts = expr_to_bytecode(expr);
            insts.push(Inst::PrintStr);
            insts
        }
    }
}

pub fn ast_to_bytecode(ast: &[Stmt]) -> Vec<Inst> {
    let mut insts = vec![];
    for stmt in ast {
        insts.extend(stmt_to_bytecode(stmt));
    }
    insts
}

#[cfg(test)]
mod tests {
    use ast::Expr;
    use token::{range, Range};

    use super::expr_to_bytecode;

    #[test]
    fn test_i32_lit_to_bytecode() {
        insta::assert_debug_snapshot!(expr_to_bytecode(&Expr::I32Lit(Range::default(), 3)));
    }

    #[test]
    fn test_str_lit_to_bytecode() {
        let bytecode = expr_to_bytecode(&Expr::StrLit(range!(0;0,0;4), "Hello".to_owned()));
        insta::assert_debug_snapshot!(bytecode);
    }

    #[test]
    fn test_addition_to_bytecode() {
        let addition = Expr::Add(
            Box::new(Expr::I32Lit(range!(0;0,0;0), 3)),
            Box::new(Expr::I32Lit(range!(0;4,0;4), 4)),
        );
        let bytecode = expr_to_bytecode(&addition);
        insta::assert_debug_snapshot!(bytecode);
    }
}
