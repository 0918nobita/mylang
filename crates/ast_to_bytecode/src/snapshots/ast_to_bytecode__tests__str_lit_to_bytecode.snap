---
source: crates/ast_to_bytecode/src/lib.rs
assertion_line: 54
expression: "expr_to_bytecode(&Expr::StrLit(range!(0 ; 0, 0 ; 4), \"Hello\".to_owned()))"

---
[
    StrConst(
        "Hello",
    ),
]
